File: Data Structures & Algorithms/maximum-product-subarray/submission-1.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        mxv = nums[0]

        n = len(nums)
        dp = [[1 for _ in range(n)] for _ in range(n)]

        for i in range(n):
            dp[i][i] = nums[i]
            mxv = max(mxv, dp[i][i])
        
        for length in range(2, n + 1):
            for i in range(n - length + 1):
                j = i+length-1
                dp[i][j] = nums[i]*dp[i+1][j]
                mxv = max(mxv, dp[i][j])

        return mxv

File: Data Structures & Algorithms/jump-game-ii/submission-1.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n, numeric_limits<int>::max());
        dp[0] = 0;
        for (int i = 0; i < nums.size(); ++i) {
            for (int j = 1; j <= min(nums[i], n-i-1); ++j) {
                dp[i+j] = min(dp[i+j], dp[i] +1);
            }
        }

        return dp[n-1];

        
    }
};
